backend/de.metas.business/src/main/java/de/metas/order/paymentschedule/OrderPaymentScheduleCreator.java

```diff
@@ -0,0 +1,208 @@
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.order.paymentschedule;
+
+import de.metas.currency.Amount;
+import de.metas.currency.CurrencyPrecision;
+import de.metas.order.IOrderBL;
+import de.metas.order.OrderId;
+import de.metas.payment.paymentterm.PaymentTerm;
+import de.metas.payment.paymentterm.PaymentTermBreak;
+import de.metas.payment.paymentterm.PaymentTermId;
+import de.metas.payment.paymentterm.PaymentTermService;
+import de.metas.payment.paymentterm.ReferenceDateType;
+import de.metas.shipping.PurchaseOrderToShipperTransportationRepository;
+import de.metas.shipping.PurchaseOrderToShipperTransportationService;
+import de.metas.util.Services;
+import de.metas.util.lang.Percent;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+import org.adempiere.exceptions.AdempiereException;
+import org.compiere.model.I_C_Order;
+import org.compiere.util.TimeUtil;
+import org.jetbrains.annotations.Nullable;
+import org.springframework.stereotype.Service;
+
+import java.sql.Timestamp;
+
+@Service
+public class OrderPaymentScheduleCreator
+{
+	public static OrderPaymentScheduleCreator newInstanceForUnitTesting()
+	{
+		final OrderPayScheduleService orderPayScheduleService = new OrderPayScheduleService(new OrderPayScheduleRepository());
+		return new OrderPaymentScheduleCreator(new PaymentTermService(), orderPayScheduleService);
+	}
+
+	private static final Timestamp PENDING_DATE = Timestamp.valueOf("9999-01-01 00:00:00");
+	private final IOrderBL orderBL = Services.get(IOrderBL.class);
+
+	private final PaymentTermService paymentTermService;
+	private final OrderPayScheduleService orderPayScheduleService;
+
+	public OrderPaymentScheduleCreator(@NonNull final PaymentTermService paymentTermService, @NonNull final OrderPayScheduleService orderPayScheduleService)
+	{
+		this.paymentTermService = paymentTermService;
+		this.orderPayScheduleService = orderPayScheduleService;
+	}
+
+	/**
+	 * Entry point to create schedules for a completed Order.
+	 */
+	public void createComplexSchedules(@NonNull final I_C_Order orderRecord)
+	{
+		final PaymentTermId paymenttermId = orderBL.getPaymentTermId(orderRecord);
+		final PaymentTerm paymentTerm = paymentTermService.getById(paymenttermId);
+
+		if (!paymentTerm.hasBreaks())
+		{
+			return;
+		}
+
+		if (orderHasLegacyPaySchedule(orderRecord))
+		{
+			throw new AdempiereException("Payment Term " + paymentTerm.getValue() +
+					" is flagged as complex and cannot coexist with legacy C_PaySchedule records.");
+		}
+
+		final Amount grandTotal = orderBL.getGrandTotal(orderRecord);
+
+		if (grandTotal.isZero())
+		{
+			return; // Nothing to schedule
+		}
+
+		final CurrencyPrecision precision = orderBL.getAmountPrecision(orderRecord);
+
+		for (final PaymentTermBreak termBreak : paymentTerm.getSortedBreaks())
+		{
+
+			createSingleOrderPaySchedule(OrderSchedulingContext.builder()
```

Comment:
feels like OrderSchedulingContext is too much.
consider inlining createSingleOrderPaySchedule and drop OrderSchedulingContext .